cpp/src/arrow/compute/exec/exec_plan.h

```diff
@@ -243,12 +244,17 @@ ExecNode* MakeSourceNode(ExecPlan* plan, std::string label,
 
 /// \brief Add a sink node which forwards to an AsyncGenerator<ExecBatch>
 ///
-/// Emitted batches will not be ordered; instead they will be tagged with the `seq` at
-/// which they were received.
+/// Emitted batches will not be ordered.
 ARROW_EXPORT
 std::function<Future<util::optional<ExecBatch>>()> MakeSinkNode(ExecNode* input,
                                                                 std::string label);
 
+/// \brief Add a sink node which forwards to a RecordBatchReader
+///
+/// Emitted batches will not be ordered.
+ARROW_EXPORT
+std::shared_ptr<RecordBatchReader> MakeSinkNodeReader(ExecNode* input, std::string label);
```

Comment:
Looks like this isn't used anywhere in this PR.

Reply:
Additionally it seems it'd be more useful to reorder them right? Though maybe that should be a separate node.

Reply:
Dataset scans can preserve ordering based on fragment/batch index, but this is only possible for graphs which don't include a pipeline breaker. I could provide `BatchGen dataset::BatchesInOrder(BatchGen)` to apply fragment/batch ordering if possible.

This is mostly here to provide for non-async consumption of an ExecPlan and to avoid requiring R to write bindings for ExecBatch

Reply:
Ah, ok, just wasn't sure of the use here. If it's for R, sounds good.